File: src/main/java/com/vk/lingvobot/factory/StateFactory.java
package com.vk.lingvobot.factory;

import com.vk.lingvobot.states.SettingsSetupState;
import com.vk.lingvobot.states.impl.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
class StateFactory implements AbstractFactory<SettingsSetupState> {

    private final PronounSetup pronounSetup;
    private final DifficultySetup difficultySetup;
    private final LessonsPerDaySetup lessonsPerDaySetup;
    private final PartOfTheDaySetup partOfTheDaySetup;
    private final EndSetup endSetup;
    private final MiscStateSetup miscStateSetup;

    @Override
    public SettingsSetupState getInstance(Integer stateIndex) {
        switch (stateIndex) {
            case 1:
                return pronounSetup;
            case 5:
                return difficultySetup;
            case 6:
                return lessonsPerDaySetup;
            case 7:
                return partOfTheDaySetup;
            case 8:
                return endSetup;
            default:
                return miscStateSetup;
        }
    }
}

File: src/main/kotlin/com/vk/lingvobot/factory/StateFactory.kt
package com.vk.lingvobot.factory

import com.vk.lingvobot.states.SettingsSetupState
import com.vk.lingvobot.states.impl.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class StateFactory @Autowired constructor(
    private val pronounSetup: PronounSetup,
    private val difficultySetup: DifficultySetup,
    private val lessonsPerDaySetup: LessonsPerDaySetup,
    private val partOfTheDaySetup: PartOfTheDaySetup,
    private val endSetup: EndSetup,
    private val miscStateSetup: MiscStateSetup
) : AbstractFactory<SettingsSetupState> {

    override fun getInstance(stateIndex: Int): SettingsSetupState {
        return when (stateIndex) {
            1 -> pronounSetup
            5 -> difficultySetup
            6 -> lessonsPerDaySetup
            7 -> partOfTheDaySetup
            8 -> endSetup
            else -> miscStateSetup
        }
    }
}